File: db/nodejs/core/file.adapter.js
import {join} from "node:path";
import {promises as fs} from "node:fs";

export class FileAdapter {
    //todo: set work folder for DB
    constructor(filename = 'samuraidb.txt') {
        // todo: join vs resolve
        this.filename = join(filename);
        this.indexFileName = 'index.txt';
    }

    async set(key, data) {
        // Сериализуем данные в JSON и создаем строку в формате "ключ,значение"
        const entry = `${key},${JSON.stringify(data)}\n`;

        // Открываем файл для получения текущего смещения
        const fileHandle = await fs.open(this.filename, 'a');
        const offset = (await fileHandle.stat()).size;

        await fs.appendFile(this.filename, entry);

        // Закрываем файл
        await fileHandle.close();

        return {offset}
    }

    async get(offset) {
        // Проверяем наличие ключа в индексе
        if (offset === undefined) {
            throw new Error('Offset  must be passed') // Если ключа нет в индексе, возвращаем null
        }

        // Открываем файл и переходим к нужному смещению
        const fileHandle = await fs.open(this.filename, 'r');

        // Читаем строку с ключом и значением
        const buffer = Buffer.alloc(1024); // Создаем буфер для чтения строки
        await fileHandle.read(buffer, 0, 1024, offset);
        const line = buffer.toString('utf-8').trim();

        // Закрываем файл
        await fileHandle.close();

        const [storedKey, storedValue] = line.split(/,(.+)/);

        return JSON.parse(storedValue);
    }

    async saveIndex(indexMap) {
        const serializedMap = JSON.stringify(Array.from(indexMap));
        await fs.writeFile(this.indexFileName, serializedMap, 'utf-8');
    }

    async readIndex() {
        try {
            const fileContent = await fs.readFile(this.indexFileName, 'utf-8');
            return new Map(JSON.parse(fileContent));
        } catch (error) {
            if (error.code === 'ENOENT') {
                // File not found, return an empty Map
                return new Map();
            }
            // Re-throw the error if it's not a "file not found" error
            throw error;
        }
    }
}

File: db/golang/internal/file-adapter/file-adapter.go
package file_adapter

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/fs"
	"log"
	"log/slog"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
)

type FileAdapter struct {
	filename      string
	indexFileName string
	mutex         sync.Mutex
}

func NewAdapter(dir string) *FileAdapter {

	filename := filepath.Join(dir, "samuraidb")
	indexFileName := filepath.Join(dir, "index.txt")

	// Проверка или создание директории
	if err := os.MkdirAll(dir, os.ModePerm); err != nil {
		log.Fatalf("Failed to create directory: %v", err)
	}

	return &FileAdapter{filename: filename, indexFileName: indexFileName}
}

func (fa *FileAdapter) Set(key string, currentSegment int, data []byte) (int64, error) {
	fa.mutex.Lock()
	defer fa.mutex.Unlock()

	entry := fmt.Sprintf("%s,%s\n", key, string(data))

	fileName := fa.getSegmentFileName(currentSegment)

	file, err := os.OpenFile(fileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return 0, err
	}
	defer file.Close()

	offset, err := file.Seek(0, io.SeekEnd)
	if err != nil {
		return 0, err
	}

	_, err = file.WriteString(entry)
	if err != nil {
		return 0, err
	}

	return offset, nil
}

func (fa *FileAdapter) Get(offset int64) (map[string]any, error) {
	if offset < 0 {
		return nil, fmt.Errorf("Offset must be passed")
	}

	file, err := os.Open(fa.filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	_, err = file.Seek(offset, io.SeekStart)
	if err != nil {
		return nil, err
	}

	buffer := make([]byte, 1024)
	n, err := file.Read(buffer)
	if err != nil {
		return nil, err
	}

	line := string(buffer[:n])
	_, storedValue, _ := parseEntry(line)

	var value map[string]any
	err = json.Unmarshal([]byte(storedValue), &value)
	if err != nil {
		return nil, err
	}
	return value, nil
}

func (fa *FileAdapter) SaveIndex(indexMap map[string]int64) error {
	fa.mutex.Lock()
	defer fa.mutex.Unlock()

	serializedMap, err := json.Marshal(indexMap)
	if err != nil {
		return fmt.Errorf("unable to serialize the data in SaveIndex: %w", err)
	}
	return os.WriteFile(fa.indexFileName, serializedMap, 0644)
}

func (fa *FileAdapter) ReadIndex() (map[string]int64, error) {
	fileContent, err := os.ReadFile(fa.indexFileName)
	if err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			return make(map[string]int64), nil
		}
		return nil, err
	}

	var index map[string]int64
	err = json.Unmarshal(fileContent, &index)
	if err != nil {
		return nil, err
	}

	return index, nil
}

func (fa *FileAdapter) GetFileSize(segmentNumber int) (int64, error) {
	fileName := fa.getSegmentFileName(segmentNumber)
	fileInfo, err := os.Stat(fileName)
	if err != nil {
		slog.Info("file data retrieval error", slog.Any("error: ", err))
		return 0, err
	}

	return fileInfo.Size(), nil
}

func (fa *FileAdapter) getSegmentFileName(segmentNumber int) string {
	fn := strings.Builder{}
	fn.WriteString(fa.filename)
	fn.WriteString("_segment_")
	fn.WriteString(strconv.Itoa(segmentNumber))
	fn.WriteString(".txt")
	return fn.String()
}

// Helper functions
func serializeData(data any) string {
	b, _ := json.Marshal(data)
	return string(b)
}

func parseEntry(line string) (string, string, error) {
	res := strings.SplitN(line, ",", 2)
	if len(res) < 2 {
		return "", "", fmt.Errorf("Invalid entry format")
	}

	return res[0], res[1], nil
}
